src/Utilities/MoistThermodynamics/src/MoistThermodynamics.jl

```diff
@@ -341,170 +603,333 @@ function saturation_shum(T, ρ, q_liq=0, q_ice=0)
-    T, converged = find_zero(roots_equation,
-                             T0, T1,
-                             args,
+    # FIXME here: need to revisit bounds for saturation adjustment to guarantee bracketing of zero.
+    T_1 = max(T_min, air_temperature(e_int, q_tot, DT(0), DT(0)))
+    T_2 = air_temperature(e_int, q_tot, DT(0), q_tot)
+    eos_root(x, ρ, q_tot, e_int) = internal_energy_sat(x, ρ, q_tot) - e_int
+    T, converged = find_zero(eos_root, T_1, T_2, args,
                              IterParams(tol_abs, iter_max),
-                             SecantMethod()
-                             )
+                             SecantMethod())
     q_liq, q_ice = phase_partitioning_eq(T, ρ, q_tot)
     return air_temperature(e_int, q_tot, q_liq, q_ice)
   end
 end
 
+
 """
-    liquid_ice_pottemp(T, p[, q_tot=0, q_liq=0, q_ice=0])
+    saturation_adjustment_q_tot_θ_liq_ice(θ_liq_ice, q_tot, ρ, p)
+
+Compute the temperature that is consistent with
+
+ - `θ_liq_ice` liquid-ice potential temperature
+ - `q_tot` total specific humidity
+ - `ρ` density
+ - `p` pressure
+
+See also [`saturation_adjustment`](@ref).
+"""
+function saturation_adjustment_q_tot_θ_liq_ice(θ_liq_ice::DT, q_tot::DT, ρ::DT, p::DT) where DT
+  T_1 = air_temperature_from_pottemp(θ_liq_ice, p) # Assume all vapor
+  qv_star = saturation_shum(T_1, ρ)
+  if (q_tot <= qv_star) # If not saturated
+    return T_1
+  else  # If saturated, iterate
+    T_2 = air_temperature_from_pottemp(θ_liq_ice, p, q_tot, DT(0), q_tot) # Assume all ice
+    function eos_root(T, args)
+      q_c   = saturation_excess(T, ρ, q_tot)
+      q_liq, q_ice = phase_partitioning_eq(T, ρ, q_tot)
+      return θ_liq_ice - liquid_ice_pottemp_sat(T, p, q_tot, q_liq, q_ice)
+    end
+    T, converged = find_zero(eos_root, T_1, T_2, Tuple(1),
+                             IterParams(DT(1e-3), 10),
+                             SecantMethod())
+    return T
+  end
+end
 
-Return the liquid-ice potential temperature, given the temperature `T`,
-pressure `p`, total specific humidity `q_tot`, liquid specific humidity `q_liq`,
-and ice specific humidity `q_ice`.
 """
-function liquid_ice_pottemp(T, p, q_tot=0, q_liq=0, q_ice=0)
+    liquid_ice_pottemp(T, p[, q_tot=0, q_liq=0, q_ice=0])
 
+The liquid-ice potential temperature where
+ - `T` temperature
+ - `p` pressure
+and, optionally,
+ - `q_tot` total specific humidity
+ - `q_liq` liquid specific humidity
+ - `q_ice` ice specific humidity
+"""
+function liquid_ice_pottemp(T::DT, p::DT, q_tot::DT=DT(0), q_liq::DT=DT(0), q_ice::DT=DT(0)) where DT
     # isobaric specific heat of moist air
     _cp_m   = cp_m(q_tot, q_liq, q_ice)
 
     # liquid-ice potential temperature, approximating latent heats
     # of phase transitions as constants
-    return dry_pottemp(T, p, q_tot, q_liq, q_ice) * exp(-(LH_v0*q_liq + LH_s0*q_ice)/(_cp_m*T))
-
+    return dry_pottemp(T, p, q_tot, q_liq, q_ice) * exp(-(DT(LH_v0)*q_liq + DT(LH_s0)*q_ice)/(_cp_m*T))
```

Comment:
Replace by linearized form, so it can easily be inverted for temperature:
`dry_pottemp(T, p, q_tot, q_liq, q_ice) * (1 - (DT(LH_v0)*q_liq + DT(LH_s0)*q_ice)/(_cp_m*T))`

Put the inverse `air_temperature_from...` function near here, and make sure it inverts this expression.